source/request_response.c

```diff
@@ -202,22 +324,7 @@ int aws_http_request_add_header(struct aws_http_request *request, struct aws_htt
     AWS_PRECONDITION(request);
     AWS_PRECONDITION(aws_byte_cursor_is_valid(&header.name) && aws_byte_cursor_is_valid(&header.value));
 
-    struct aws_http_header_impl header_impl;
-    int err = s_header_impl_init(&header_impl, &header, request->allocator);
-    if (err) {
-        goto error;
-    }
-
-    err = aws_array_list_push_back(&request->headers, &header_impl);
-    if (err) {
-        goto error;
-    }
-
-    return AWS_OP_SUCCESS;
-
-error:
-    s_header_impl_clean_up(&header_impl);
-    return AWS_OP_ERR;
+    return s_header_block_add_header(&request->headers, request->allocator, header);
```

Comment:
for all these functions, maybe move the PRECONDITIONS (except AWS_PRECONDITION(request);
) to the header-block function?